hummingbot/connector/derivative/bybit_perpetual/bybit_perpetual_user_stream_data_source.py

```diff
@@ -0,0 +1,135 @@
+from hummingbot.logger import HummingbotLogger
+
+
+class BybitPerpetualUserStreamDataSource(UserStreamTrackerDataSource):
+    _logger: Optional[HummingbotLogger] = None
+
+    @classmethod
+    def logger(cls) -> HummingbotLogger:
+        if cls._logger is None:
+            cls._logger = logging.getLogger(__name__)
+        return cls._logger
+
+    def __init__(self, auth_assistant: BybitPerpetualAuth, session: Optional[aiohttp.ClientSession] = None, domain: Optional[str] = None):
+        super().__init__()
+        self._auth_assistant: BybitPerpetualAuth = auth_assistant
+        self._last_recv_time: float = 0
+        self._domain = domain
+        self._session = session
+
+    @property
+    def last_recv_time(self) -> float:
+        return self._last_recv_time
+
+    async def _create_websocket_connection(self) -> BybitPerpetualWebSocketAdaptor:
+        """
+        Initialize WebSocket client for UserStreamDataSource
+        """
+        try:
+            if self._session is None:
+                self._session = aiohttp.ClientSession()
+            ws = await self._session.ws_connect(bybit_perpetual_utils.wss_url(self._domain))
+            return BybitPerpetualWebSocketAdaptor(websocket=ws)
+        except asyncio.CancelledError:
+            raise
+        except Exception as ex:
+            self.logger().network(f"Unexpected error occurred during {CONSTANTS.EXCHANGE_NAME} WebSocket Connection "
+                                  f"({ex})")
+            raise
+
+    async def _authenticate(self, ws: BybitPerpetualWebSocketAdaptor):
+        """
+        Authenticates user to websocket
+        """
+        self.logger().info("_authenticate")
+        try:
+            auth_payload: Dict[str, Any] = self._auth_assistant.get_ws_auth_payload()
+            await ws.authenticate(auth_payload)
+            auth_resp = await ws.receive_str()
+            auth_payload: Dict[str, Any] = ws.payload_from_raw_message(auth_resp)
+
+            if auth_payload["success"] is not True or not auth_payload["request"] or not auth_payload["request"]["op"] or auth_payload["request"]["op"] != "auth":
+                self.logger().error(f"Response: {auth_payload}", exc_info=True)
+                raise Exception("Could not authenticate websocket connection with Bybit Perpetual")
+
+        except asyncio.CancelledError:
+            raise
+        except Exception as ex:
+            self.logger().error(f"Error occurred when authenticating to user stream ({ex})",
+                                exc_info=True)
+            raise
+
+    async def _subscribe_to_events(self, ws: BybitPerpetualWebSocketAdaptor):
+        """
+        Subscribes to User Account Events
+        """
+        try:
+            await ws.subscribe_to_positions()
+            await ws.subscribe_to_orders()
+            await ws.subscribe_to_stop_orders()
+
+            auth_resp = await ws.receive_str()
```

Comment:
This is not correct for Bybit. The server could start sending events down the websocket even before sending the response to the subscription request.